### PR TITLE
Added option to skip committing atoms to database after inference

### PR DESCRIPTION
**Pull request by vihanggodbole**

### CODE REVIEW

File: psl-cli/src/main/java/org/linqs/psl/cli/Launcher.java [hunk 1]
```diff
@@ -198,7 +199,9 @@ private Database runInference(Model model, DataStore dataStore, Set<StandardPred
             outputGroundRules(inferenceApplication.getGroundRuleStore(), path, false);
         }
 
-        inferenceApplication.inference();
+        Boolean skipAtomCommit = parsedOptions.hasOption(CommandLineLoader.OPTION_SKIP_ATOM_COMMIT_LONG) ? true : false;
```
Comment (eriq-augustine) [inline psl-cli/src/main/java/org/linqs/psl/cli/Launcher.java]: Prefer POD over Object.

File: psl-cli/src/main/java/org/linqs/psl/cli/Launcher.java [hunk 2]
```diff
@@ -198,7 +199,9 @@ private Database runInference(Model model, DataStore dataStore, Set<StandardPred
...
+
+        inferenceApplication.inference(skipAtomCommit);
```
Comment (eriq-augustine) [inline psl-cli/src/main/java/org/linqs/psl/cli/Launcher.java]: Avoid auto-boxing.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java [hunk 3]
```diff
@@ -31,6 +32,8 @@
 import org.linqs.psl.reasoner.term.TermGenerator;
 import org.linqs.psl.reasoner.term.TermStore;
 import org.linqs.psl.util.Reflection;
+import org.slf4j.Logger;
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java:35]: Third-party imports should be its own section: This Project (PSL) -> Third-party -> Standard Library.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java [hunk 4]
```diff
@@ -74,7 +77,9 @@
     protected GroundRuleStore groundRuleStore;
     protected TermStore termStore;
     protected TermGenerator termGenerator;
-    protected AtomManager atomManager;
+    protected PersistedAtomManager atomManager;
+
+    private static final Logger log = LoggerFactory.getLogger(LazyMPEInference.class);
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java]: !! Wrong Class object!

Make this the very first statement in the class.
Static members should generally be before non-static ones.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java [hunk 5]
```diff
@@ -109,14 +114,39 @@ protected void initialize() {
...
+
     /**
      * Minimizes the total weighted incompatibility of the GroundAtoms in the Database
      * according to the Model and commits the updated truth values back to the Database.
      *
      * The RandomVariableAtoms to be inferred are those persisted in the Database when this method is called.
      * All RandomVariableAtoms which the Model might access must be persisted in the Database.
      */
-    public abstract void inference();
+    public void inference(boolean skipAtomCommit) {
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java]: Swap `skipAtomCommit` for `commitAtoms`.
I think it is more intuitive for people to pass `true` when they want something to be done.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java [hunk 6]
```diff
@@ -109,14 +114,39 @@ protected void initialize() {
...
+        log.info("Beginning inference.");
+        internalInference();
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java:153]: Log right after this line.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java [hunk 7]
```diff
@@ -109,14 +114,39 @@ protected void initialize() {
...
+
+        // Commits the RandomVariableAtoms back to the Database.
+        if (!skipAtomCommit) {
+            log.info("Inference complete. Writing results to Database.");
+            atomManager.commitPersistedAtoms();
+            log.info("Results committed to database.");
+        } else {
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java]: No `else`, the log will be moved up.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java [hunk 8]
```diff
@@ -109,14 +114,39 @@ protected void initialize() {
...
      * All RandomVariableAtoms which the Model might access must be persisted in the Database.
      */
-    public abstract void inference();
+    public void inference(boolean skipAtomCommit) {
+        log.info("Beginning inference.");
+        internalInference();
+
+        // Commits the RandomVariableAtoms back to the Database.
+        if (!skipAtomCommit) {
+            log.info("Inference complete. Writing results to Database.");
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java]: Remove "Inference complete" from log.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java [hunk 9]
```diff
@@ -74,7 +77,9 @@
     protected GroundRuleStore groundRuleStore;
     protected TermStore termStore;
     protected TermGenerator termGenerator;
-    protected AtomManager atomManager;
+    protected PersistedAtomManager atomManager;
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java:82]: Almost all the IA's construct the PAM in the same way (except lazy).
Let's factor it out to a method that lazy can override.
Something like `initializeAtomManager()`.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java [hunk 10]
```diff
@@ -109,14 +114,39 @@ protected void initialize() {
...
+     */
+    public void inference() {
+        inference(false);
+    }
+
     /**
      * Minimizes the total weighted incompatibility of the GroundAtoms in the Database
      * according to the Model and commits the updated truth values back to the Database.
      *
      * The RandomVariableAtoms to be inferred are those persisted in the Database when this method is called.
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java]: Redundant lines.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java [hunk 11]
```diff
@@ -109,14 +114,39 @@ protected void initialize() {
      */
     protected abstract void completeInitialize();
 
+    /**
+     * Minimizes the total weighted incompatibility of the GroundAtoms in the Database
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java]: Just make this comment something short like:
"Alias for inference() with committing atoms."

File: psl-core/src/main/java/org/linqs/psl/application/inference/LazyMPEInference.java [hunk 12]
```diff
@@ -74,9 +74,9 @@ protected void completeInitialize() {
     }
 
     @Override
-    public void inference() {
-        inference(model.getRules(), reasoner, groundRuleStore, termStore, termGenerator,
-                (LazyAtomManager)atomManager, maxRounds);
+    protected void internalInference() {
+        inference(model.getRules(), reasoner, groundRuleStore, termStore, termGenerator, (LazyAtomManager) atomManager,
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/LazyMPEInference.java]: Extra space after cast.

File: psl-core/src/main/java/org/linqs/psl/application/inference/SGDStreamingInference.java [hunk 13]
```diff
@@ -51,14 +51,8 @@ protected void completeInitialize() {
     }
 
     @Override
-    public void inference() {
-        log.info("Beginning inference.");
+    protected void internalInference() {
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/SGDStreamingInference.java]: Look like all (except lazy) `internalInference()` are now the same.
Instead, you can probably provide a default implementation of `internalInference()` that lazy overrides.

File: psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java [hunk 14]
```diff
@@ -195,4 +197,68 @@ public void testLogicalTautologyTrivial() {
         mpe.close();
         inferDB.close();
     }
+
+    /**
+    * Make sure that skipAtomCommit flag from CLI works and no atom is persisted in the database.
+    */
+    @Test
+    public void testSkipAtomCommit() {
+        // Arrange
```
Comment (eriq-augustine) [inline psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java]: Uninformative comment.

File: psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java [hunk 15]
```diff
@@ -195,4 +197,68 @@ public void testLogicalTautologyTrivial() {
         mpe.close();
         inferDB.close();
     }
+
+    /**
+    * Make sure that skipAtomCommit flag from CLI works and no atom is persisted in the database.
```
Comment (eriq-augustine) [inline psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java]: You are not testing the CI here.

File: psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java [hunk 16]
```diff
@@ -195,4 +197,68 @@ public void testLogicalTautologyTrivial() {
...
+    */
+    @Test
+    public void testSkipAtomCommit() {
+        // Arrange
+        TestModel.ModelInformation info = TestModel.getModel();
+        Set<StandardPredicate> toClose = new HashSet<StandardPredicate>();
+        Database inferDB = info.dataStore.getDatabase(info.targetPartition, toClose, info.observationPartition);
+        MPEInference mpe = new MPEInference(info.model, inferDB);
```
Comment (eriq-augustine) [inline psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java]: This variable should just be an IA (the object can still be `MPEInference`).
Always make variables as general as they can be, e.g. you don't want to pass around an `ArrayList`, you want a `List` instead.
This is VERY important, talk to me if you don't understand.

File: psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java [hunk 17]
```diff
@@ -195,4 +197,68 @@ public void testLogicalTautologyTrivial() {
...
+                
+        float preInferenceTotalValue = 0.0f;
+        for (RandomVariableAtom atom : inferDB.getAllGroundRandomVariableAtoms(info.predicates.get("Friends"))) {
+            preInferenceTotalValue += atom.getValue();
+        }
+
+        // Act
```
Comment (eriq-augustine) [inline psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java]: Uninformative.

File: psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java [hunk 18]
```diff
@@ -195,4 +197,68 @@ public void testLogicalTautologyTrivial() {
...
+        mpe.inference(true);
+        
+        mpe.close();
+        inferDB.close();
+        
+        inferDB = info.dataStore.getDatabase(info.targetPartition, toClose, info.observationPartition);
```
Comment (eriq-augustine) [inline psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java:225]: Comment this line and explain what is happening.
Remember, other people may look at this test later.

File: psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java [hunk 19]
```diff
@@ -195,4 +197,68 @@ public void testLogicalTautologyTrivial() {
...
+        float postInferenceTotalValue = 0.0f;
+        for (RandomVariableAtom atom : inferDB.getAllGroundRandomVariableAtoms(info.predicates.get("Friends"))) {
+            postInferenceTotalValue += atom.getValue();
+        }
+
+        // Assert
```
Comment (eriq-augustine) [inline psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java]: Uninformative.

File: psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java [hunk 20]
```diff
@@ -195,4 +197,68 @@ public void testLogicalTautologyTrivial() {
...
+        assertEquals(preInferenceTotalValue, postInferenceTotalValue, 0.001f);
+    }
+
+    /**
+    * Make sure that not setting the skipAtomCommit flag from CLI works and all atoms are persisted in the database.
```
Comment (eriq-augustine) [inline psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java]: All the same comments from the previous test.

File: psl-parser/src/main/java/org/linqs/psl/parser/CommandLineLoader.java [hunk 21]
```diff
@@ -326,6 +327,9 @@ private static Options setupOptions() {
                 .valueSeparator('=')
                 .build());
 
+        newOptions.addOption(Option.builder(OPTION_SKIP_ATOM_COMMIT_LONG).longOpt(OPTION_SKIP_ATOM_COMMIT_LONG)
```
Comment (eriq-augustine) [inline psl-parser/src/main/java/org/linqs/psl/parser/CommandLineLoader.java]: See `OPTION_OUTPUT_SATISFACTION_LONG` on a long option without a short option.

**Review by eriq-augustine — CHANGES_REQUESTED** [review]
Getting there, but you made some sloppy mistakes.

File: psl-cli/src/main/java/org/linqs/psl/cli/Launcher.java [hunk 22]
```diff
@@ -198,7 +199,9 @@ private Database runInference(Model model, DataStore dataStore, Set<StandardPred
             outputGroundRules(inferenceApplication.getGroundRuleStore(), path, false);
         }
 
-        inferenceApplication.inference();
+        boolean commitAtoms = parsedOptions.hasOption(CommandLineLoader.OPTION_SKIP_ATOM_COMMIT_LONG) ? false : true;
```
Comment (eriq-augustine) [inline psl-cli/src/main/java/org/linqs/psl/cli/Launcher.java]: In general ternary operators are usually the wrong choice.
(I love them, but there is usually a better way.)

Just negate `parsedOptions.hasOption()`.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java [hunk 23]
```diff
@@ -21,6 +21,7 @@
 import org.linqs.psl.config.Config;
 import org.linqs.psl.database.Database;
 import org.linqs.psl.database.atom.AtomManager;
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java]: Unused import.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java [hunk 24]
```diff
@@ -32,10 +33,16 @@
...
 import org.linqs.psl.util.Reflection;
 
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
 import java.lang.reflect.Constructor;
 import java.lang.reflect.InvocationTargetException;
 
 public abstract class InferenceApplication implements ModelApplication {
+
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java]: Extra newline.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java [hunk 25]
```diff
@@ -74,12 +81,16 @@
...
     protected TermStore termStore;
     protected TermGenerator termGenerator;
-    protected AtomManager atomManager;
+    protected PersistedAtomManager atomManager;
 
     public InferenceApplication(Model model, Database db) {
         this.model = model;
         this.db = db;
 
+        log.debug("Creating persisted atom mannager.");
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java]: You can do this in `initialize()`?

File: psl-core/src/main/java/org/linqs/psl/application/inference/DCDStreamingInference.java [hunk 26]
```diff
@@ -34,31 +38,28 @@ public DCDStreamingInference(Model model, Database db) {
...
+    protected TermGenerator createTermGenerator() {
+        return null;
+    }
 
-        // Commits the RandomVariableAtoms back to the Database,
-        ((PersistedAtomManager)atomManager).commitPersistedAtoms();
-        log.info("Results committed to database.");
+    @Override
+    protected void completeInitialize() {
+        termStore.ensureVariableCapacity(atomManager.getCachedRVACount());
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/DCDStreamingInference.java]: Lets do this in `InferenceApplication`, make it common for all children.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java (hunk 10 above)
Comment (eriq-augustine) [reply psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java]: This is not resolved.

File: psl-core/src/main/java/org/linqs/psl/application/inference/LazyMPEInference.java [hunk 27]
```diff
@@ -20,6 +20,7 @@
 import org.linqs.psl.config.Config;
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/LazyMPEInference.java:20]: Unused import.

File: psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java (hunk 18 above)
Comment (eriq-augustine) [reply psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java:225]: Unresolved comment.

File: psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java [hunk 28]
```diff
@@ -195,4 +197,68 @@ public void testLogicalTautologyTrivial() {
...
+        float preInferenceTotalValue = 0.0f;
+        for (RandomVariableAtom atom : inferDB.getAllGroundRandomVariableAtoms(info.predicates.get("Friends"))) {
+            preInferenceTotalValue += atom.getValue();
+        }
+
+        mpe.inference(true);
+        
+        mpe.close();
+        
+        // The database should be closed after inference, before reading in values again, to clear cache
```
Comment (eriq-augustine) [inline psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java]: "clear the cache".

Also if comments are getting long, try to break them on clauses.

**Review by eriq-augustine — CHANGES_REQUESTED** [review]
Getting closer.
Let's try and close this out today.

File: psl-core/src/main/java/org/linqs/psl/application/inference/LazyMPEInference.java (hunk 27 above)
Comment (vihanggodbole) [reply psl-core/src/main/java/org/linqs/psl/application/inference/LazyMPEInference.java:20]: It is used for the constants in the constructor.

File: psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java [hunk 29]
```diff
@@ -88,35 +93,80 @@ public InferenceApplication(Model model, Database db) {
...
+
     /**
      * Complete the initialization process.
      * Most of the infrastructure will have been constructued.
      * The child is responsible for constructing the AtomManager
      * and populating the ground rule store.
      */
-    protected abstract void completeInitialize();
+    protected void completeInitialize() {
+        termStore.ensureVariableCapacity(atomManager.getCachedRVACount());
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/InferenceApplication.java]: Move this to `initialize()`.

File: psl-core/src/main/java/org/linqs/psl/application/inference/LazyMPEInference.java [hunk 30]
```diff
@@ -16,10 +16,10 @@
  * limitations under the License.
  */
 package org.linqs.psl.application.inference;
-
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/LazyMPEInference.java]: Missing space.

File: psl-core/src/main/java/org/linqs/psl/application/inference/MPEInference.java [hunk 31]
```diff
@@ -45,10 +45,6 @@ public MPEInference(Model model, Database db) {
 
     @Override
     protected void completeInitialize() {
-        log.debug("Creating persisted atom mannager.");
-        atomManager = new PersistedAtomManager(db);
-        log.trace("Atom manager initialization complete.");
-
         log.info("Grounding out model.");
         int groundCount = Grounding.groundAll(model, atomManager, groundRuleStore);
         log.info("Grounding complete.");
```
Comment (eriq-augustine) [inline psl-core/src/main/java/org/linqs/psl/application/inference/MPEInference.java:50]: `ensureVariableCapacity` got moved.

File: psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java [hunk 32]
```diff
@@ -195,4 +197,70 @@ public void testLogicalTautologyTrivial() {
...
+
+        mpe.inference(true);
+        
+        mpe.close();
+        
+        // The database should be closed after inference to clear the cache before reading in values again.
+        // This ensures that the values returned won't be from the cache.
+        inferDB.close();
+        
+        // Create a new instance of database with an empty cache.
```
Comment (eriq-augustine) [inline psl-core/src/test/java/org/linqs/psl/application/inference/MPEInferenceTest.java]: "a database".

(Same below.)

**Review by eriq-augustine — CHANGES_REQUESTED** [review]
Getting closer.

**Review by eriq-augustine — APPROVED** [review]

**Pull request merged**